pkg/framework/proxies.go

```diff
@@ -0,0 +1,249 @@
+									SubPath:   "squid.conf",
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{},
+						},
+					},
+					NodeSelector: map[string]string{
+						"proxy-host": "",
+					},
+					HostNetwork: true,
+				},
+			},
+			Strategy: appsv1.DeploymentStrategy{
+				Type: "Recreate",
+			},
+			Paused: false,
+		},
+	}
+	c.Create(context.Background(), deployment)
+	IsDeploymentAvailable(c, deployment.Name, "default")
+	return deployment, nil
+}
+
+// DestroyClusterProxy destroys the HTTP proxy and associated configmap
+func DestroyClusterProxy(c runtimeclient.Client) error {
+	configMap := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "squid-conf",
+			Namespace: "default",
+		},
+	}
+	c.Delete(context.Background(), configMap)
+
+	return DeleteDeployment(c, &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "squid-proxy",
+			Namespace: "default",
+		},
+	})
+}
+
+// WaitForProxyInjectionSync waits for the deployment to sync with the state of the cluster-proxy
+func WaitForProxyInjectionSync(c runtimeclient.Client, name, namespace string, shouldBePresent bool) bool {
+	if err := wait.PollImmediate(RetryShort, WaitLong, func() (bool, error) {
+		deployment, err := GetDeployment(c, name, namespace)
+		if err != nil {
+			return false, nil
+		}
+		hasHttpProxy := false
+		hasHttpsProxy := false
+		hasNoProxy := false
+		for _, container := range deployment.Spec.Template.Spec.Containers {
+			for _, envVar := range container.Env {
+				switch envVar.Name {
+				case "NO_PROXY":
+					hasNoProxy = true
+				case "HTTPS_PROXY":
+					hasHttpsProxy = true
+				case "HTTP_PROXY":
+					hasHttpProxy = true
+				}
+			}
+		}
+		return (hasHttpProxy &&
+			hasHttpsProxy &&
+			hasNoProxy) == shouldBePresent, nil
+	}); err != nil {
+		klog.Errorf("Error checking isDeploymentAvailable: %v", err)
+		return false
+	}
+	return true
+}
+
+// GetClusterProxy fetches the global cluster proxy object.
+func GetClusterProxy(c runtimeclient.Client) (*configv1.Proxy, error) {
+	proxy := &configv1.Proxy{}
+	proxyName := runtimeclient.ObjectKey{
+		Name: GlobalInfrastuctureName,
+	}
+
+	if err := c.Get(context.Background(), proxyName, proxy); err != nil {
+		return nil, err
+	}
+
+	return proxy, nil
+}
+
+// SetClusterProxy Configures the cluster-wide proxy
+func SetClusterProxy(c runtimeclient.Client, updatedProxy configv1.Proxy) (*configv1.Proxy, error) {
+	proxy, err := GetClusterProxy(c)
+	if err != nil {
+		return nil, err
+	}
+	if updatedProxy.Spec.HTTPProxy != "" {
+		proxy.Spec.HTTPProxy = updatedProxy.Spec.HTTPProxy
+	} else {
+		proxy.Spec.HTTPProxy = ""
+	}
+	if updatedProxy.Spec.HTTPSProxy != "" {
+		proxy.Spec.HTTPSProxy = updatedProxy.Spec.HTTPSProxy
+	} else {
+		proxy.Spec.HTTPSProxy = ""
+	}
+	if updatedProxy.Spec.NoProxy != "" {
+		proxy.Spec.NoProxy = updatedProxy.Spec.NoProxy
+	} else {
+		proxy.Spec.NoProxy = ""
+	}
```

Comment:
All of these can be condensed to 
```suggestion
	proxy.Spec.HTTPProxy = updatedProxy.Spec.HTTPProxy
	proxy.Spec.HTTPSProxy = updatedProxy.Spec.HTTPSProxy
	proxy.Spec.NoProxy = updatedProxy.Spec.NoProxy
```
But do we really need to do this? Would we not want to just `c.Update(context.Background(), updatedProxy)`? Do we even need this wrapper function?

Reply:
very good call.  when I wrote that I thought those variables inited to nil when undefined, which they don't.  I'll make that change.